### PR TITLE
[CPP Graph] ChatGLM Enabling and ChatGLM-2 Issues Fix

### PR DESCRIPTION
**Pull request by Zhenzhong1**
## Type of Change

Feature add and Issues Fix

## Description

New models add:
* [ChatGLM-6B](https://huggingface.co/THUDM/chatglm-6b)

Add extension-test

  - https://github.com/intel-innersource/frameworks.ai.lpot.lpot-validation/pull/501
  - Fully extension tests: https://inteltf-jenk.sh.intel.com/blue/organizations/jenkins/ITREX-cpp-graph-extension/detail/ITREX-cpp-graph-extension/84/pipeline/
  - ChatGLM & ChatGLM-2 extension tests https://inteltf-jenk.sh.intel.com/blue/organizations/jenkins/ITREX-cpp-graph-extension/detail/ITREX-cpp-graph-extension/82/pipeline

Performance Thanks @violetch24 
01628641-b9dd-456c-bc12-c95b2291f0d3

Accuracy  Thanks @violetch24
bf76c442-615f-4fe1-88d3-a53b2b1ff6d6

9549333e-6161-49b6-948c-a1fceb6595c6

Releated https://github.com/intel/intel-extension-for-transformers/pull/210 issues fixed:

- [x]  Get n_ctx by ctx->model.hparams.n_ctx.
        The rope requires the size of prompts but not the fixed model.haprams.n_ctx. No need to chage.
- [x] add parameters list bos_token_id / eos_token_id / pad_token_id / sep_token_id to voca.
        These parameters has been added to model.vocab.
- [x] multi-query-attention(MQA) double check for ChatGLM-2.
        Use n_head_kv instead of multi_query_group_num may cause accuracy issues.

## Expected Behavior & Potential Risk

N/A.

## How has this PR been tested?

Manual and CI tests.

## Dependency Change?

N/A.

### CODE REVIEW

**Comment by Zhenzhong1** [comment]
https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/ITREX-cpp-graph-extension/78/

**Comment by kevinintel** [comment]
please provide acc and perf with ffn

**Comment by Zhenzhong1** [comment]
> please provide acc and perf with ffn

I listed in the introduction of this PR.

File: intel_extension_for_transformers/llm/runtime/graph/application/common.cpp [hunk 1]
```diff
@@ -1113,3 +1113,63 @@ bool console_readline(console_state& con_st, std::string& line) {
...
+    }
+    // blank tokens
+    {
+        static const std::regex pattern(R"(<\|blank_(\d+)\|>)");
+        output = regex_replace(output, pattern,
+                               [](const std::smatch &sm) { return std::string(std::stoi(sm[1].str()), ' '); });
+    }
+
+    return output;
+}
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/application/common.cpp:1175]: remove the EOF

File: intel_extension_for_transformers/llm/runtime/graph/application/main_run.cpp [hunk 2]
```diff
@@ -654,7 +666,16 @@ int main(int argc, char** argv) {
     }
 
     // end of text token
-    if (!embd.empty() && embd.back() == model_token_eos()) {
+    if (params.model_arch == MODEL_CHATGLM) {
+      if (!embd.empty() && embd.back() == 130005) {
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_run.cpp:670]: I'd prefer to set 130005 as a static explicit variable

**Comment by Zhenzhong1** [comment]
> 
This link is the quick rebuild link of the ChatGLM.

I am debuging the env of the ChatGLM extension-test. The last successful extension-test I will list in the intro of this PR.

**Comment by Zhenzhong1** [comment]
> > https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/ITREX-cpp-graph-extension/78/
> 
> failed?

This link is the quick rebuild link of the ChatGLM.

I am debuging the env of the ChatGLM extension-test. The last successful extension-test I will list in the intro of this PR.

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp [hunk 3]
```diff
@@ -106,128 +112,139 @@ static bool chatglm_model_eval_internal(model_context& lctx, const model_token*
...
       cur = ne_add(ctx0, cur, model.layers[il].attn[1]);
+      //cur = ne_add(ctx0, ne_repeat(ctx0, model.layers[il].attn[1], cur), cur);  // ChatGLM-1 [qlen, 3 * hidden] , ChatGLM-2 [qlen, hidden + 2 * kv_hidden]
 
-      ne_tensor* query_layer = ne_view_3d(ctx0, cur, head_size, n_head, N, 3 * head_size * ne_element_size(cur),
-                                          cur->nb[1], 0);  // [qlen, 3 * hidden]
+      // ChatGLM-1
+      ne_tensor *query_layer = ne_view_3d(ctx0, cur, head_size, n_head, N, 3 * head_size * ne_element_size(cur), cur->nb[1], 0); // [qlen, 3 * hidden]
 
       ne_set_name(query_layer, "query_layer");
+      //query_layer = ne_rope_inplace(ctx0, query_layer, n_past, rope_dim, 0);  // ChatGLM-2
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp:137]: I will prefer default parameters as we talked before
```
NE_API struct ne_tensor* ne_rope_inplace(struct ne_context* ctx, struct ne_tensor* a, int n_past, int n_dims, int mode, int n_ctx = 0);
```

or you can define a new function like
```
NE_API struct ne_tensor* ne_rope_inplace_real(struct ne_context* ctx, struct ne_tensor* a, int n_past, int n_dims, int mode, int n_ctx);
```

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp [hunk 4]
```diff
@@ -106,128 +112,139 @@ static bool chatglm_model_eval_internal(model_context& lctx, const model_token*
...
 
-      ne_tensor* attn_probs = ne_soft_max_inplace(ctx0, attn_scores);  // [heads, qlen, klen]
-
-      ne_tensor* context_layer = ne_mul_mat(ctx0, value_layer, attn_probs);  // [heads, qlen, head_size]
-
+      ne_tensor *context_layer = ne_mul_mat(ctx0, value_layer, attn_probs); // [heads, qlen, head_size]
+      
       context_layer = ne_cont(ctx0, ne_permute(ctx0, context_layer, 0, 2, 1, 3));
-
+
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp:202]: format

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp [hunk 5]
```diff
@@ -266,6 +286,13 @@ static bool chatglm_model_eval_internal(model_context& lctx, const model_token*
       logits_out.resize(n_vocab);
       memcpy(logits_out.data(), (float*)ne_get_data(inpL), sizeof(float) * n_vocab);
     }
+    
+    //printf("  lctx.logits = %f \n", lctx.logits[0]);
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp:290]: remove

File: intel_extension_for_transformers/llm/runtime/graph/scripts/run_llm.py [hunk 6]
```diff
@@ -17,6 +17,7 @@
 import argparse
 from typing import List, Optional
 import subprocess
+from transformers import AutoTokenizer
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/scripts/run_llm.py:20]: better prepare sth like ```requirements.txt``` since you add new dependency

**Review by airMeng — CHANGES_REQUESTED** [review]

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp (hunk 3 above)
Comment (airMeng) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp:137]: revert changes in gptj and llama with whatever option you choose

File: intel_extension_for_transformers/llm/runtime/graph/application/common.cpp (hunk 1 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/application/common.cpp:1176]: Fixed

File: intel_extension_for_transformers/llm/runtime/graph/application/main_run.cpp (hunk 2 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_run.cpp]: fixed

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp (hunk 4 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp]: clang-format done

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp (hunk 5 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp]: done

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp [hunk 7]
```diff
@@ -67,15 +64,24 @@ static bool chatglm_model_eval_internal(model_context& lctx, const model_token*
   const int n_embd = hparams.n_embd;
   const int n_layer = hparams.n_layer;
   const int n_ctx = hparams.n_ctx;
-
+
```
[the remark this answers is not in the document]
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp]: clang-format done

File: intel_extension_for_transformers/llm/runtime/graph/scripts/run_llm.py (hunk 6 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/scripts/run_llm.py:20]: Done. graph/requirements.txt has included the transformer already.

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp [hunk 8]
```diff
@@ -67,15 +64,24 @@ static bool chatglm_model_eval_internal(model_context& lctx, const model_token*
...
+  //const int tokens_size = tokens.size();
+  //printf("n_tokens = %d, n_past = %d \n", n_tokens, n_past);
+
   const int n_head = hparams.n_head;
   const int n_vocab = hparams.n_vocab;
   const int n_rot = n_embd / n_head / 2;
+  //const int head_size = n_embd / n_head;
+  //const int rope_dim = head_size / 2;
+  // const int mqa_scale = n_head / hparams.multi_query_group_num;
+  // const int num_kv_heads = hparams.multi_query_group_num;
```
[the remark this answers is not in the document]
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp]: I have tried to use n_head_kv instead of multi_query_group_num. It caused accuracy issues.
d2597d39-e945-4420-9d13-935892781139

These two parameters are differerent. Some models need them both. They are not the totally same.

[the remark this answers is not in the document]
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp]: n_head_kv is also used in some unifed places. @zhentaoyu @DDEle 
kvsize0

[the remark this answers is not in the document]
Comment (zhentaoyu) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp]: falcon_7b: ext test: https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/ITREX-cpp-graph-extension/80/
falcon_7b `n_head_kv` = 1. Let's see what its kv cache memory size firstly.

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_types.h
Comment (DDEle) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_types.h:1]: https://github.com/intel/intel-extension-for-transformers/pull/210#discussion_r1319230559

How is this issue going?

Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_types.h:1]: Replied

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (zhentaoyu) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp]: It will be investigated in the next pr. Resolve it for successful running and unified arch.

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp (hunk 3 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm.cpp]: I've rebuilt fully extended tests on the branch of this PR. 
https://inteltf-jenk.sh.intel.com/blue/organizations/jenkins/ITREX-cpp-graph-extension/detail/ITREX-cpp-graph-extension/84/pipeline

As we discussed, adding the dfault parameters for this c file may introduce more unnecessary code because native C does not support default parameters for functions. 

Can we keep the same as ggml rope kernel, which means no change?

**Review by airMeng — APPROVED** [review]
please start another extension test done

**Comment by airMeng** [comment]
@zhenwei-intel please review the perf

**Review by a32543254 — APPROVED** [review]
LGTM

**Comment by a32543254** [comment]
Consider to run chat glm 13b

**Comment by zhenwei-intel** [comment]
> @zhenwei-intel please review the perf

LGTM
